astropy/timeseries/downsample.py

```diff
@@ -61,45 +81,93 @@ def aggregate_downsample(time_series, *, time_bin_size=None, time_bin_start=None
+        # In this case, we require time_bin_size to be specified
+        if time_bin_size.isscalar:
+            if n_bins is None:
+                bin_size_sec = time_bin_size.to_value(u.s)
+                n_bins = int(np.ceil(time_duration/bin_size_sec))
+
+    # Check for overlapping bins, and warn if they are present
+    if time_bin_end is not None:
+        if not time_bin_end.isscalar and not time_bin_start.isscalar:
+            if np.any(time_bin_start[1:] < time_bin_end[:-1]):
+                warnings.warn("Overlapping bins should be avoided since they "
+                              "can lead to double-counting of data during binning.",
+                              AstropyUserWarning)
+
+    binned = BinnedTimeSeries(time_bin_size=time_bin_size,
+                              time_bin_start=time_bin_start,
+                              time_bin_end=time_bin_end,
+                              n_bins=n_bins)
 
     if aggregate_func is None:
         aggregate_func = np.nanmean
 
-    # Determine the bins
-    relative_bins_sec = np.cumsum(np.hstack([0, np.repeat(bin_size_sec, n_bins)]))
-    bins = time_bin_start + relative_bins_sec * u.s
+    # Start and end times of the binned timeseries
+    bin_start = binned.time_bin_start
+    bin_end = binned.time_bin_end
+
+    # Assign `n_bins` since it is needed later
+    if n_bins is None:
+        n_bins = len(bin_start)
+
+    # Find the relative time since the start time, in seconds
+    relative_time_sec = (sorted.time - bin_start[0]).sec
+    # Duration of binned timeseries in seconds
+    bin_duration = (bin_end[-1] - bin_start[0]).sec
+
+    bin_start_sec = (bin_start - bin_start[0]).sec
+    bin_end_sec = (bin_end - bin_start[0]).sec
 
     # Find the subset of the table that is inside the bins
-    keep = ((relative_time_sec >= relative_bins_sec[0]) &
-            (relative_time_sec < relative_bins_sec[-1]))
-    subset = sorted[keep]
+    keep = ((relative_time_sec >= 0) &
+            (relative_time_sec <= bin_duration))
+
+    # Find out indices to be removed because of uncontiguous bins
+    for ind in range(n_bins-1):
+        delete_indices = np.where(np.logical_and(relative_time_sec > bin_end_sec[ind],
+                                                 relative_time_sec < bin_start_sec[ind+1]))
+        keep[delete_indices] = False
 
-    # Figure out which bin each row falls in - the -1 is because items
-    # falling in the first bins will have index 1 but we want that to be 0
-    indices = np.searchsorted(relative_bins_sec, relative_time_sec[keep]) - 1
-    # Add back the first time.
-    indices[relative_time_sec[keep] == relative_bins_sec[0]] = 0
+    subset = sorted[keep]
 
-    # Create new binned time series
-    binned = BinnedTimeSeries(time_bin_start=bins[:-1], time_bin_end=bins[-1])
+    # Figure out which bin each row falls in by sorting with respect
+    # to the bin end times
+    indices = np.searchsorted(bin_end_sec, relative_time_sec[keep])
```

Comment:
Nice! Thanks for the suggestion. I was looking into this part earlier since I realized that's where the issue was arising. However, couldn't think of something on the fly that would work for all different cases without having to handle each one separately. 

Maybe `np.ndenumerate` might work for for assigning _indices_. I will try that.

I'll also try to think of an elegant solution to handle overlapping bins separately.